cadc-gms/src/main/java/org/opencadc/auth/StandardIdentityManager.java

```diff
@@ -335,43 +334,53 @@ private void validateOidcAccessToken(Subject s) {
-                        s.getPrincipals().add(hp);
-
+                if (credentials != null) {
+                    // Bearer tokens are handled specially by validating with the Identity Provider
+                    if (AuthenticationUtil.CHALLENGE_TYPE_BEARER.equalsIgnoreCase(challengeType)) {
+                        try {
+                            HttpGet get = new HttpGet(u, true);
+                            get.setRequestProperty("authorization", raw.getHeaderValue());
+                            get.prepare();
+
+                            InputStream istream = get.getInputStream();
+                            String str = StringUtil.readFromInputStream(istream, "UTF-8");
+                            JSONObject json = new JSONObject(str);
+                            String sub = json.getString("sub");
+                            String username = json.getString("preferred_username");
+                            // TODO: register an X509 DN with IAM and see if I can get it back here
+
+                            OpenIdPrincipal oip = new OpenIdPrincipal(oidcClient.getIssuerURL(), sub);
+                            HttpPrincipal hp = new HttpPrincipal(username);
+
+                            s.getPrincipals().remove(raw);
+                            s.getPrincipals().add(oip);
+                            s.getPrincipals().add(hp);
+
+                            AuthorizationToken authToken = new AuthorizationToken(challengeType, credentials, oidcDomains, oidcScope);
+                            s.getPublicCredentials().add(authToken);
+                        } catch (NotAuthenticatedException ex) {
+                            JSONObject json = new JSONObject(ex.getMessage());
+                            String error = json.getString("error");
+                            String details = json.getString("error_description");
+                            // details usually includes the invalid access token: truncate
+                            StringBuilder sb = new StringBuilder(error);
+                            sb.append(" reason: ");
+                            int max = Math.min(details.length(), 32);
+                            sb.append(details.subSequence(0, max));
+                            if (max < details.length()) {
+                                sb.append("...");
+                            }
+                            throw new NotAuthenticatedException(challengeType, NotAuthenticatedException.AuthError.INVALID_TOKEN, sb.toString());
+                        } catch (Exception ex) {
+                            throw new NotAuthenticatedException(challengeType, NotAuthenticatedException.AuthError.INVALID_TOKEN, ex.getMessage(), ex);
+                        }
+                    } else {
+                        // Otherwise it's a custom token, so we just create an AuthorizationToken and let whatever
+                        // application is looking for it handle it.
```

Comment:
No. I thought it was clear that this IM impl should just leave the other ATP alone. If some other code does not deal with it, then authentication is rejected: that is the intended behaviour.